docs/Fleet management Design.md

```diff
@@ -0,0 +1,163 @@
+    // Name is the name of a ClusterClaim resource on managed cluster. It's a well known
+    // or customized name to identify the claim.
+    Name string `json:"name,omitempty"`
+    
+    // Value is a claim-dependent string
+    Value string `json:"value,omitempty"`
+}
+
+// ResourceList defines a map for the quantity of different resources, the definition
+// matches the ResourceList defined in k8s.io/api/core/v1.
+type ResourceList map[ResourceName]resource.Quantity
+
+```
+
+### Design details
+Here is how we handle the following functionalities in the OCM in our new controller. The overall idea is that 
+a `managedCluster` CR represents a member cluster. A cluster joins the fleet when a corresponding `managedCluster` CR is
+successfully applied to the hub cluster. Similarly, a member cluster leaves the hub cluster when its corresponding 
+`managedCluster` CR is removed from the cluster. 
+
+#### Join a member cluster to the fleet
+The managedCluster will look for a secret named `name of the cluster`-`kubeconfig`. 
+For example, the controller that reconciles a `managedCluster` called "member-cluster-A" will look for
+a secret named `member-cluster-A-kubeconfig`. The secret contains the entire kubeconfig data. The controller will then 
+create a namespace of the same name if it does not already exist.
+
+There is another way to provide the credentials of member clusters to the hub cluster. There is a `ClientConfig`
+field in the `managedCluster` definition, described above, that is designed to contain the access info.
+However, the upstream OCM project currently is not utilizing it so that it does not contain the necessary fields to keep
+the certificate and key data. We may propose to add the credential data fields to the CRD in the future so that it saves
+
+#### Collect the cluster status and claims
+In the current OCM, the Klusterlet controller updates the status of a `managedCluster` which includes the following fields
+
+* Capacity
+* Allocatable
+* Version (ManagedClusterVersion)
+* ClusterClaims
+
+The first three fields contains information that one can get from kube client by querying all the nodes periodically.
+The last one has to be provided by the member cluster. There is already a customer object named `ClusterClaim` that contains
+the required information.  Therefore, we propose that the `managedCluster` controller to just list the `ClusterClaim` objects in the 
+member cluster to fill this part.
+
+#### Work with `work` API
+The `work` API is the way how OCM manage "applications". Its controller is running in the member cluster in the current pull
+model. To make it work with push model, we need to implement it with the `managedCluster` controller. Here is how it works
+
+* The `managedCluster` controller reconciler keeps an in-memory map of member cluster -> manifestwork controller
+* for each new member cluster, it creates a new manifestwork controller that watches all the `ManifestWork` objects in the 
```

Comment:
The previous experience on kubefed is it will be pretty resource intensive if a monolithic controller on hub to apply resources to a bunch of spokes. The work agent itself can be run on the hub, so it is possible you have multiple work agents running on hub for each spoke. Would you callout the cons and pros for each approach?

Reply:
yes, the push model does not scale well. Let me clarify that we don't intend to use the push model to replace the pull model. It serves as a lightweight complement.